### PR TITLE
Drop federation transmission queues during a significant remote outage.

### LINKED ISSUE #7828: in-memory federation transaction transmission queues build up indefinitely for offline servers
when a server is offline, any events that we would have sent to it stack up in memory indefinitely.

See, for example, this graph for matrix.org:

image

Apart from the memory usage concerns, this means that if a server goes offline for 1000 events, then when it comes back we will try to send it all those 1000 events in turn, rather than just sending it the most recent ones in a room.

### PR DESCRIPTION
**Pull request by reivilibre**
Fixes #7828. Depends on #2528.

Signed-off-by: Olivier Wilkinson (reivilibre) <olivier@librepush.net>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by Half-Shot** [comment]
This sounds like you woulf drop transactions on a intermittent failure too though? E.g. if s server is upgraded, or network troubles, or you are matrix.org. Emptying a transaction too soon might cause even more netsplits?

**Comment by jcgruenhage** [comment]
I agree with @Half-Shot here, this sounds a bit too overzealous. Another concern: While servers can easily backfill missing PDUs in a room, is the same the case for missed EDUs? If we randomly stop sending EDUs to other servers, this is going to result in messages which are unable to decrypt, because the to_device messages get lost, right?

**Comment by richvdh** [comment]
> This sounds like you would drop transactions on a intermittent failure too though? E.g. if s server is upgraded, or network troubles, or you are matrix.org. 

Well, we only do this after trying to send a given transaction 11 times (those retries happen at a lower level, in the `MatrixFederationHttpClient` class). If you assume each request times out (after 60s), that means we try to send each transaction for *over 20 minutes* before giving up on it. There's a limit to what we can plausibly call an "intermittent failure" before we should give up and move on.

> While servers can easily backfill missing PDUs in a room, is the same the case for missed EDUs? 

In the general case, no. However, the EDUs that are being dropped here are those that we can afford to drop (specifically, only typing notifications are being dropped here). Other EDUs such as to_device messages are queued with a different mechanism. (Indeed, if the only record we had of to_device messages was these in-memory queues, we'd have big problems whenever the federation sender process got restarted.)

**Comment by Half-Shot** [comment]
>Well, we only do this after trying to send a given transaction 11 times (those retries happen at a lower level, in the MatrixFederationHttpClient class). If you assume each request times out (after 60s), that means we try to send each transaction for over 20 minutes before giving up on it. There's a limit to what we can plausibly call an "intermittent failure" before we should give up and move on.

Ah, that's what I was missing then. I thought this was a drop immediately after a failure. 20 minutes sounds very reasonable.

File: synapse/federation/sender/per_destination_queue.py [hunk 1]
```diff
@@ -337,6 +337,11 @@ async def _transaction_transmission_loop(self) -> None:
                     (e.retry_last_ts + e.retry_interval) / 1000.0
                 ),
             )
+
+            # don't let the pending event queues stack up
+            self._pending_pdus = []
+            self._pending_edus = []
+            self._pending_edus_keyed = {}
```
Comment (erikjohnston) [inline synapse/federation/sender/per_destination_queue.py:344]: Do we not also need to clear read receipts?

**Comment by erikjohnston** [comment]
FTR I don't think we can assume that intermittent failures will lead to timeouts, if synapse dies or is down for maintenance then we'll likely get a connection refused response very quickly, which I think takes the total time down to sub 10 minutes. That is easily plausible for a deploy-gone-wrong or down-for-brief maintenance; jki.re certainly sometimes goes down for that long semi regularly.

I'm a little bit hesitant with being this aggressive in general, and certainly wouldn't want to merge this until we have landed #2528 as this will just make that issue so much worse. 

Personally I'd probably look at the `retry_interval` in the exception, and only throw away if its greater than an hour or something. Possibly much longer before we flush the PDU queue until we have fixed #2528.

**Comment by auscompgeek** [comment]
What if the queues were just trimmed down to the latest 10 or so events?

**Comment by reivilibre** [comment]
I think I agree that it seems like this should only be put in after a solution to #2528 (which will also solve the case of losing touch with other servers after a 'natural' loss of transmission queues because of an outage, upgrade or other restart).

However, does this seem reasonable now, assuming that mechanism arrives?

File: changelog.d/7864.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix in-memory federation transmission queues building up indefinitely.
```
Comment (clokep) [inline changelog.d/7864.bugfix:1]: What's this mean to an end-user? Is there a behavior that is fixed with this?

**Comment by richvdh** [comment]
I'd kinda like us to get this merged rather than have it hanging around for weeks in purgatory. @erikjohnston are you strongly opposed to it?

**Comment by clokep** [comment]
This could likely use a merge of develop at this point too, just to ensure churn hasn't caused issues.

**Comment by erikjohnston** [comment]
This seems fine to me

File: changelog.d/7864.bugfix (hunk 2 above)
Comment (reivilibre) [reply changelog.d/7864.bugfix:1]: How about

> Stop leaking memory when remote servers have been unreachable for a long time.

?

Comment (clokep) [reply changelog.d/7864.bugfix:1]: Maybe:

> Fix a memory leak by limiting the number of messages queued for a remote server that has been unreachable for a long time.

**Pull request merged**